keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -360,10 +373,20 @@ Type VolumeGroupSnapshotStatus struct {
         // +optional
         Error *VolumeSnapshotError
 
-        // List of volume snapshot refs
-        // The max number of snapshots in the group is 100
-        // +optional
-        VolumeSnapshotRefList []core_v1.ObjectReference
+	// VolumeSnapshotRefList is the list of PVC and VolumeSnapshot pairs that
+	// is part of this group snapshot.
+	// The maximum number of allowed snapshots in the group is 100.
+	// +optional
+	PVCVolumeSnapshotRefList []PVCVolumeSnapshotPair
+}
+
+// PVCVolumeSnapshotPair defines a pair of a PVC reference and a Volume Snapshot Reference
+type PVCVolumeSnapshotPair struct {
+	// PersistentVolumeClaimRef is a reference to the PVC this pair is referring to
+	PersistentVolumeClaimRef core_v1.LocalObjectReference
+
+	// VolumeSnapshotRef is a reference to the VolumeSnapshot this pair is referring to
+	VolumeSnapshotRef core_v1.LocalObjectReference
```

Comment:
Starting a new "thread" about this flow:

> 1. User creates a VolumeGroupSnapshot API object.
> 2. A CSI driver creates a volume group snapshot in one gRPC call. As a result, it returns a group snapshot id and a list of snapshots (with source volume id and snapshot id).
> 3. The group controller logic will then create VolumeSnapshotContent that points to the existing snapshot id and create VolumeSnapshot to bound to the pre-provisioned snapshot.

It's perhaps too late to challenge that, but we created external-snapshotter intentionally with very restricted permissions. ([v6.0.0 RBACs](https://github.com/kubernetes-csi/external-snapshotter/blob/83d0e0914f696b6f686f568ce0c2e99d1cbe3269/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L33-L41))

Now it can create almost anything snapshot related in the whole cluster.

This workflow would keep all permissions in snapshot-controller and IMO would be also more obvious to users:

1. User creates a VolumeGroupSnapshot API object.
2. The snapshot-controller creates VolumeSnapshots and VolumeSnapshotContents with non empty volumeGroupSnapshotName - those would be ignored by the external-snapshotter.
3. The snapshot-controller finally creates VolumeGroupSnapshotContent.
4. Presence of VolumeGroupSnapshotContent (or some field there that says the controller is done) would be a signal for the external-snapshotter to collect all volume handles from VolumeGroupSnapshotContents that are part of that group, call the CSI driver and file VolumeSnapshotContents and VolumeGroupSnapshotContent.

All using the APIs that we already have, without any pairs anywhere.

Reply:
I agree that the workflow outlined above has benefits as you mentioned.

The rationale for the existing workflow is as follows:

* Hold off the creation of individual VolumeSnapshot and VolumeSnapshotContent API objects as much as possible and only create them when snapshots are already created successfully on the storage backend. This can prevent multiple individual VolumeSnapshot and VolumeSnapshotContent API objects from staying pending for a long time and prevent those entries from flooding the logs (if the creation of a volume group snapshot takes a long time in a busy system). It will be easier to troubleshoot if creating a volume group snapshot fails or times out.